Update grammar and translation code, regenerate data

cutiful opened the pull request:
This PR updates ES5 and ES2015 formal data and adds ES2016-ES2022. To achieve that, it also updates the Jison grammar specification and the translation code.

Fixes #3.

cutiful committed:
Pull latest ESTree

cutiful committed:
Update spec filenames, add newer specs

cutiful committed:
Update package-lock.json

cutiful committed:
Update grammar to allow enum and interface inheritance extensions

cutiful committed:
Extend enums correctly

cutiful committed:
Fix extending bases

RReverser commented:
Heh that code could clearly be modernised by now. The changes look good from a quick look.

RReverser commented on src/index.js:
```diff
@@ -27,11 +27,19 @@ function resolveExtends(extension, base) {
 			let baseItem = base[name];
+			if (item.base)
+				for (const base of item.base)
+					if (baseItem.base.indexOf(base) === -1)
```
```suggestion
					if (!baseItem.base.includes(base))

```

RReverser commented on src/index.js:
```diff
@@ -27,11 +27,19 @@ function resolveExtends(extension, base) {
 			});
+		} else if (item.kind === 'enum' && name in base) {
+			result[name] = base[name];
+			result[name].values.push(...item.values);
```
Hm, wait, doesn't this modify the base object? With `merge` used in other branches we create a copy, but this one seems to retain original object by reference.

cutiful replied:
Oh, that's right — I didn't realize the base object is important, since I write the file immediately after the merge (and I didn't see the potential race condition). I'll fix that.

RReverser reviewed (commented):
(left some comments)

cutiful committed:
Rewrite loop as recursive function to avoid potential race conditions

cutiful committed:
Regenerate formal data

cutiful commented:
I rewrote the extension resolution code to avoid modifying base in both interfaces and enums. I didn't realize that the ES2020 spec included previously possible values in the LogicalOperator extension; using Set removed the duplicates. Outside that generated data hasn't changed.

I also rewrote the loop into a recursive function, to make sure race conditions don't happen. It would be better to use async/await for that purpose, but I didn't want to introduce those into code that uses bluebird. (Yeah, it could be modernized, but that takes time, and I doubt many people look here for the code and not the jsons 😅)

(Sorry for a lot of force-pushes, here's the [link](https://github.com/estree/formal/compare/3acfc1f9ae9ec0c24d316b8585151e5c5b1fbba3..484059abc93f505dfe725cef9c3e8f4a8848206b) to compare the original PR to the latest version.)

RReverser commented:
> It would be better to use async/await for that purpose, but I didn't want to introduce those into code that uses bluebird.

I'm definitely up for removing bluebird, it was necessary back when this was written, but not anymore. Same for async-await :)

RReverser commented on src/index.js:
```diff
@@ -27,10 +27,18 @@ function resolveExtends(extension, base) {
 			result[name] = merge(baseItem, {
-				props: merge(baseItem.props, item.props)
+				props: merge(baseItem.props, item.props),
+				base: [...new Set(baseItem.base.concat(item.base || []))]
```
Maybe extract this into a helper `mergeArr` for clarity that would do `new Set` + `.concat` internally for both usages.

cutiful committed:
Extract array merging into a helper function

cutiful commented:
> Maybe extract this into a helper mergeArr for clarity that would do new Set + .concat internally for both usages.

Done!

> I'm definitely up for removing bluebird, it was necessary back when this was written, but not anymore. Same for async-await :)

I think that belongs to a separate issue.

RReverser commented on src/index.js:
```diff
@@ -10,10 +10,14 @@ installSourceMaps();
 }
 
+function mergeArr(...arrays) {
+	return [...new Set([].concat(...arrays))];
```
Little simpler:

```suggestion
	return [...new Set(arrays.flat())];
```

cutiful replied:
Oh, cool!

cutiful committed:
Compile updated sources

RReverser commented:
> I think that belongs to a separate issue.

Done on master with some other improvements (e.g. typechecking via jsdoc).
